File: backend/app/Services/AI/AIRouter.php
<?php

namespace App\Services\AI;

use Illuminate\Support\Facades\Log;

class AIRouter
{
    protected ProviderRegistry $registry;
    protected CircuitBreaker $circuitBreaker;
    protected int $maxAttempts;
    protected int $maxRetries;

    public function __construct(ProviderRegistry $registry, CircuitBreaker $circuitBreaker)
    {
        $this->registry = $registry;
        $this->circuitBreaker = $circuitBreaker;
        $this->maxAttempts = (int) env('AI_MAX_TOTAL_ATTEMPTS', 6);
        $this->maxRetries = (int) env('AI_MAX_RETRIES_PER_PROVIDER', 1);
    }

    public function chat(array $messages, array $options = []): array
    {
        $capabilities = $options['capabilities'] ?? [];
        $order = $this->registry->getOrder();
        $adapters = $this->registry->getAdapters();

        $attemptCount = 0;

        foreach ($order as $providerId) {
            if (!isset($adapters[$providerId])) continue;
            
            $adapter = $adapters[$providerId];
            
            if (!$adapter->isCapable($capabilities)) {
                Log::debug("AI Router: Skipping {$providerId}, lacks requested capabilities.");
                continue;
            }

            $state = $this->circuitBreaker->getState($providerId);
            if ($state === CircuitBreaker::STATE_OPEN) {
                Log::debug("AI Router: Skipping {$providerId}, circuit is OPEN.");
                continue;
            }

            // Optional: Dynamic model discovery validate on startup. 
            // In a real app we might cache the valid models per provider. 
            // For now, assume model configured is valid and handle failures via catch.

            $retries = 0;
            
            while ($retries <= $this->maxRetries && $attemptCount < $this->maxAttempts) {
                $attemptCount++;
                try {
                    $response = $adapter->chat($messages, $options);

                    // Output structured check
                    if (in_array('JSON', $capabilities) || in_array('STRUCTURED', $capabilities)) {
                        $parsed = json_decode($response['content'], true);
                        if (json_last_error() !== JSON_ERROR_NONE) {
                            throw new \Exception("Invalid JSON output from provider");
                        }
                    }

                    $this->circuitBreaker->recordSuccess($providerId);
                    return $response;

                } catch (\Throwable $e) {
                    $errorInfo = $adapter->normalizeError($e);
                    
                    if (str_contains($e->getMessage(), "Invalid JSON output")) {
                        Log::warning("AI Router: {$providerId} returned invalid JSON. Attempt {$retries}");
                        if ($retries < $this->maxRetries) {
                            $retries++;
                            // Repair hint
                            $messages[] = ['role' => 'user', 'content' => 'The previous response was not valid JSON. Please reply ONLY with valid JSON.'];
                            continue;
                        }
                        $this->circuitBreaker->recordFailure($providerId);
                        break; // Failover
                    }

                    Log::warning("AI Router: {$providerId} failed: " . $e->getMessage());

                    if ($errorInfo['type'] === 'RETRY' && $retries < $this->maxRetries) {
                        $retries++;
                        sleep(1 * $retries); // simple backoff
                        continue;
                    }

                    // Failover
                    $this->circuitBreaker->recordFailure($providerId);
                    break; 
                }
            }

            if ($attemptCount >= $this->maxAttempts) {
                break;
            }
        }

        throw new \Exception("AI Router: Exhausted all available providers or reached max attempts.");
    }
}

File: worker/new/services/ai/ai_router.py
import os
import json
import asyncio
import logging
from typing import List, Dict, Any
from .provider_registry import registry
from .circuit_breaker import CircuitBreaker, STATE_OPEN
from .adapters import create_adapter

logger = logging.getLogger(__name__)

class AIRouter:
    def __init__(self):
        self.circuit_breaker = CircuitBreaker()
        self.max_attempts = int(os.getenv('AI_MAX_TOTAL_ATTEMPTS', '6'))
        self.max_retries = int(os.getenv('AI_MAX_RETRIES_PER_PROVIDER', '1'))
        
        self.adapters = {}
        for pid, config in registry.get_providers().items():
            adapter = create_adapter(pid, config)
            if adapter:
                self.adapters[pid] = adapter

    async def chat(self, messages: List[dict], options: dict = None) -> dict:
        options = options or {}
        capabilities = options.get('capabilities', [])
        
        attempt_count = 0
        
        for provider_id in registry.get_order():
            adapter = self.adapters.get(provider_id)
            if not adapter:
                continue

            if not adapter.is_capable(capabilities):
                logger.debug(f"AI Router: Skipping {provider_id}, lacks requested capabilities.")
                continue

            state = await self.circuit_breaker.get_state(provider_id)
            if state == STATE_OPEN:
                logger.debug(f"AI Router: Skipping {provider_id}, circuit is OPEN.")
                continue

            retries = 0
            
            # Create a copy of messages in case we need to append repair hints
            current_messages = list(messages)
            
            while retries <= self.max_retries and attempt_count < self.max_attempts:
                attempt_count += 1
                try:
                    response = await adapter.chat(current_messages, options)

                    # Output structured check
                    caps_upper = [c.upper() for c in capabilities]
                    if 'JSON' in caps_upper or 'STRUCTURED' in caps_upper:
                        try:
                            parsed = json.loads(response['content'])
                        except json.JSONDecodeError:
                            raise ValueError("Invalid JSON output from provider")

                    await self.circuit_breaker.record_success(provider_id)
                    return response

                except Exception as e:
                    error_info = adapter.normalize_error(e)
                    
                    if "Invalid JSON output" in str(e):
                        logger.warning(f"AI Router: {provider_id} returned invalid JSON. Attempt {retries}")
                        if retries < self.max_retries:
                            retries += 1
                            # Repair hint
                            current_messages.append({'role': 'user', 'content': 'The previous response was not valid JSON. Please reply ONLY with valid JSON.'})
                            continue
                        
                        await self.circuit_breaker.record_failure(provider_id)
                        break # Failover to next provider

                    logger.warning(f"AI Router: {provider_id} failed: {error_info.get('message', str(e))}")

                    if error_info['type'] == 'RETRY' and retries < self.max_retries:
                        retries += 1
                        await asyncio.sleep(1 * retries) # simple backoff
                        continue

                    # Failover
                    await self.circuit_breaker.record_failure(provider_id)
                    break 

            if attempt_count >= self.max_attempts:
                break

        raise Exception("AI Router: Exhausted all available providers or reached max attempts.")

# Global router instance
router = AIRouter()
